content/source/assets/javascripts/application.js

```diff
@@ -7,9 +7,45 @@
 
 //= require analytics
 
+// In navigation sidebars, hide most sub-lists and reveal any that contain the
+// current page. The a.current-page class is added during build by
+// layouts/inner.erb.
 document.addEventListener("turbolinks:load", function() {
     "use strict";
     var docsSidebar = $("#docs-sidebar ul.nav.docs-sidenav");
     docsSidebar.find("ul.nav").addClass("nav-hidden");
     docsSidebar.find("li").has("a.current-page").addClass("active");
 });
+
+// On docs/content pages, add a hierarchical quick nav menu if there are any
+// H2/H3/H4 headers.
+document.addEventListener("turbolinks:load", function() {
+    "use strict";
+    var headers = $('#inner').find('h2, h3, h4');
+    if (headers.length > 0) {
+        $("#inner #inner-quicknav").html(
+            '<span id="inner-quicknav-trigger">Page Quick Nav<svg width="9" height="5" xmlns="http://www.w3.org/2000/svg"><path d="M8.811 1.067a.612.612 0 0 0 0-.884.655.655 0 0 0-.908 0L4.5 3.491 1.097.183a.655.655 0 0 0-.909 0 .615.615 0 0 0 0 .884l3.857 3.75a.655.655 0 0 0 .91 0l3.856-3.75z" fill-rule="evenodd"/></svg></span><ul class="dropdown"></ul>'
+        );
+        var quickNavTrigger = $('#inner-quicknav #inner-quicknav-trigger');
+        var quickNav = $('#inner-quicknav > ul.dropdown');
+        headers.each( function(index, element) {
+            var level = element.nodeName.toLowerCase();
+            var header_text = $(element).text();
+            var header_id = $(element).attr('id');
+            quickNav.append(`<li class="level-${level}"><a href="#${header_id}">${header_text}</a></li>`);
+        });
+        quickNavTrigger.on('click', function(e) {
+            $(this).siblings('ul').toggleClass('active');
+            e.stopPropagation();
+        });
+        quickNav.on('click', function(e) {
+            e.stopPropagation();
+        });
+        quickNav.find('li a').on('click', function() {
+            quickNav.removeClass('active');
+        });
+        $('body').on('click', function() {
+            quickNav.removeClass('active');
+        });
```

Comment:
Just a general question - since turbolinks prevents a full reload, does every page navigation transition continue stacking more and more of these event listeners? If so, maybe should we remove the event listeners somehow - didn't see a "before route change" event [here](https://github.com/turbolinks/turbolinks#full-list-of-events)

Reply:
That's a great question. Lemme see if I can find out using the inspector tools.

I can't seem to do anything that causes doubled event listeners. After reading more, it looks like: 

- [It at least doesn't cache event listeners.](https://github.com/turbolinks/turbolinks#understanding-caching) So they shouldn't get duplicated when someone navigates history w/ back and forward.
- Turbolinks prevents a full reload, but it replaces the whole `<body>` element, right? That _should_ cause it to discard any elements that have event listeners attached.

So, it seems like we should be good. But it's also possible I'm not seeing prod-like behavior in my preview... I'll build a complete copy of the site and test with that, as well.

Reply:
Sure enough, even with a pre-built copy nothing seems to cause duplicated event listeners. The replacement of the entire `<body>` element and the fact that it doesn't cache event listeners for history navigation takes care of it, as expected.

Reply:
Hmm, thinking more about this: Turbolinks WOULD probably cache the `<span>` and `<ul>` that we add to the div, since those are new DOM elements. But since we're adding them with `.html("etc")`, it's actually idempotent -- `.html()` replaces the current contents of that div. 

Which means that on back/forward navigation, we aren't messing anything up, but we ARE doing a small amount of extra work by rebuilding the quick-nav. 

I could throw that into some `if` statements so reattach the event listeners but don't rebuild the elements. But that might be premature optimization, since it doesn't really feel bad as-is.

Reply:
Great investigation! It seems like, [so long as when the element is removed no references to it remain](https://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory), we have no potential for a memory leak here. I assume that turbolinks isn't holding on to internal references in this manner, and if it did, someone would have called it out a while ago, so we're probably good here.